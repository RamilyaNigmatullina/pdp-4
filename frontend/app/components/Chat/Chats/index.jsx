import React from 'react';
import PropTypes from 'prop-types';

class Chats extends React.Component {
  renderChat = (chat) => (
    <div key={chat.id}>
      <a href="#" className="chat-link" onClick={(e) => this.props.onClick(e, chat.id)}>
        { chat.interlocutor.full_name }
      </a>
    </div>
  )

  render() {
    return (
      <div className="chats-list">
        { this.props.chats.map((chat) => this.renderChat(chat)) }
      </div>
    );
  }
}

Chats.propTypes = {
  chats: PropTypes.array.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default Chats;
