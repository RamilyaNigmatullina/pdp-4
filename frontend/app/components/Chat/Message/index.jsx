import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

class Message extends React.Component {
  render() {
    const { avatarUrl, message, isCurrentUserMessage } = this.props;
    const messageClasses = classNames('message', {
      'message--current-user': isCurrentUserMessage,
      'message--interlocutor': !isCurrentUserMessage,
    });
    return (
      <div className={messageClasses}>
        { !isCurrentUserMessage && <img className="rounded-circle" src={avatarUrl} width="32" height="32" /> }
        <span className="message__text">{message.text}</span>
        { isCurrentUserMessage && <img className="rounded-circle" src={avatarUrl} width="32" height="32" /> }
      </div>
    );
  }
}

Message.propTypes = {
  avatarUrl: PropTypes.string,
  isCurrentUserMessage: PropTypes.bool,
  message: PropTypes.object,
};

export default Message;
