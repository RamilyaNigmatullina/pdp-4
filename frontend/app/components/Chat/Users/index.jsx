import React from 'react';
import PropTypes from 'prop-types';

class Users extends React.Component {
  render() {
    return (
      <div>
      </div>
    );
  }
}

Users.propTypes = {
  users: PropTypes.array.isRequired,
};

export default Users;
