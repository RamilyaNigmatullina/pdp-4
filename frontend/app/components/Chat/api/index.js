import axios from 'axios';

const MESSAGE_FIELDS = ['id', 'text'];

export const createMessage = (chatId, message) => axios
  .post(`/api/v1/chats/${chatId}/messages`, { ...message, fields: MESSAGE_FIELDS })
  .then((response) => response.data);
