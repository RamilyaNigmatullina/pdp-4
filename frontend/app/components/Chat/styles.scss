@import "~bootstrap/scss/functions";
@import "~bootstrap/scss/variables";
@import "~bootstrap/scss/mixins/breakpoints";

body.chats {
  height: 100vh;
  overflow: hidden;
}

.chat {
  display: flex;
  flex-direction: column;
  height: calc(100vh - 64px);

  @include media-breakpoint-down(md) {
    height: calc(100vh - 56px);
  }
}
