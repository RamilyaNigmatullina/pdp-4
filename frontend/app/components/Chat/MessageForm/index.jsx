import React from 'react';
import PropTypes from 'prop-types';
import $ from 'jquery';
import 'form-serializer';
import { createMessage } from '../api/index';

class MessageForm extends React.Component {
  constructor(props) {
    super(props);

    this.textInput = React.createRef();
  }

  renderTextarea() {
    return (
      <div className="form-group text message-form__text">
        <textarea
          ref={this.textInput}
          className="form-control"
          rows="1"
          placeholder="Type text..."
          name="message[text]"
        />
      </div>
    );
  }

  render() {
    return (
      <form
        className="simple_form message-form"
        method="post"
        action={`/api/v1/chats/${this.props.chatId}/messages`}
        onSubmit={this.handleSubmit}
      >

        { this.renderTextarea() }
        <button className="btn btn-primary message-form__button">Send</button>
      </form>
    );
  }

  handleSubmit = (event) => {
    event.preventDefault();

    const formData = $(event.target).serializeObject();
    createMessage(this.props.chatId, formData)
      .then(() => {
        this.textInput.current.value = '';
      });
  }
}

MessageForm.propTypes = {
  chatId: PropTypes.number.isRequired,
};

export default MessageForm;
