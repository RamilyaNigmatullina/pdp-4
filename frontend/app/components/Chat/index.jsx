import React from 'react';
import PropTypes from 'prop-types';
import Message from './Message';
import MessageForm from './MessageForm';

class Chat extends React.Component {
  renderMessages() {
    const { interlocutor, currentUser, messages } = this.props;

    return messages.map((message) => {
      const isCurrentUserMessage = currentUser.id === message.sender_id;
      const avatarUrl = isCurrentUserMessage ? currentUser.avatar : interlocutor.avatar;

      return <Message
        avatarUrl={avatarUrl}
        isCurrentUserMessage={isCurrentUserMessage}
        key={message.id}
        message={message}
      />;
    });
  }

  render() {
    return (
      <div className="chat">
        {this.props.interlocutor.full_name}

        <div className="messages">
          {this.renderMessages()}
        </div>

        { <MessageForm chatId={this.props.chat.id} /> }
      </div>
    );
  }
}

Chat.propTypes = {
  interlocutor: PropTypes.object,
  currentUser: PropTypes.object,
  messages: PropTypes.array,
  chat: PropTypes.object,
};

export default Chat;
