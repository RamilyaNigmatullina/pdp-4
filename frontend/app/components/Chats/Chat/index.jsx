import React from 'react';
import PropTypes from 'prop-types';
import consumer from 'helpers/consumer';
import InfiniteScrolling from 'components/InfiniteScrolling';
import Message from './Message';
import MessageForm from './MessageForm';
import Interlocutor from './Interlocutor';
import { fetchMessages } from '../api/index';

class Chat extends React.Component {
  state = {
    isLastPage: false,
    messages: [],
    page: 1,
  };

  componentDidMount() {
    this.setState({
      subscription: this.subscribeToChannel(),
    });
  }

  componentDidUpdate(prevProps) {
    if (prevProps.chat.id !== this.props.chat.id) {
      this.setState({
        isLastPage: false,
        messages: [],
        page: 1,
        subscription: this.subscribeToChannel(),
      });
    }
  }

  subscribeToChannel = () => {
    const { subscription } = this.state;

    if (subscription) subscription.unsubscribe();

    return consumer.subscriptions.create({
      channel: 'ChatChannel',
      chat_id: this.props.chat.id,
    }, {
      received: this.handleReceived,
    });
  }

  renderMessage = (message) => {
    const { currentUser, chat } = this.props;
    const isCurrentUserMessage = this.isCurrentUserMessage(message);
    const avatarUrl = isCurrentUserMessage ? currentUser.avatar : chat.interlocutor.avatar;

    return (
      <Message
        avatarUrl={avatarUrl}
        isCurrentUserMessage={isCurrentUserMessage}
        key={message.id}
        message={message}
    />);
  }

  render() {
    const { messages, isLastPage } = this.state;
    const { chat } = this.props;

    return (
      <>
        { <Interlocutor interlocutor={chat.interlocutor} /> }

        <div className="messages">
          { <InfiniteScrolling
              isLastPage={isLastPage}
              items={messages}
              loadItems={this.handleLoadMessages}
              renderItem={this.renderMessage}
            /> }
        </div>

        { <MessageForm chatId={chat.id} /> }
      </>
    );
  }

  handleReceived = (message) => {
    if (this.isCurrentUserMessage(message)) this.handleMessageReceived(message);

    this.setState(({ messages }) => ({ messages: [message, ...messages] }), () => {
      const messageElement = document.getElementById(`message-${message.id}`);
      messageElement.scrollIntoView({ behavior: 'smooth' });
    });
  }

  handleMessageReceived = (message) => {
    const chat = { ...this.props.chat, last_message: message, unread_messages_count: 0 };
    const { onMessageReceived } = this.props;

    onMessageReceived(chat);
  }

  handleLoadMessages = () => {
    fetchMessages(this.props.chat.id, this.state.page)
      .then((data) => {
        this.setState(({ messages, page }) => ({
          isLastPage: !data.length,
          messages: [...messages, ...data],
          page: page + 1,
        }));
      });
  }

  isCurrentUserMessage = (message) => this.props.currentUser.id === message.sender_id
}

Chat.propTypes = {
  chat: PropTypes.object.isRequired,
  currentUser: PropTypes.object.isRequired,
  onMessageReceived: PropTypes.func.isRequired,
};

export default Chat;
