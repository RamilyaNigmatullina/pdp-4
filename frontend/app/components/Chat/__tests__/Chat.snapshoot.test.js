import React from 'react';
import { shallow } from 'enzyme';
import Chat from '../index';

describe('Chat', () => {
  let props;

  const render = () => shallow(<Chat {...props} />);

  beforeEach(() => {
    props = {
      interlocutor: {
        id: 1,
        full_name: 'John Smith',
        avatar: '/john_smith_avatar.jpg',
      },
      currentUser: {
        id: 2,
        full_name: 'Kevin Doe',
        avatar: '/kevin_doe_avatar.jpg',
      },
      messages: [
        {
          id: 1,
          text: 'Hello, Kevin!',
          sender_id: 1,
        },
        {
          id: 2,
          text: 'Hello, John!',
          sender_id: 2,
        },
        {
          id: 1,
          text: 'How are you?',
          sender_id: 1,
        },
      ],
      chat: { id: 1 },
    };
  });

  it('renders chat', () => {
    const component = render();

    expect(component).toMatchSnapshot();
  });
});
