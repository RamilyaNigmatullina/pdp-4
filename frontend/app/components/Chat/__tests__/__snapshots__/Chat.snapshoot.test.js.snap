// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat renders chat 1`] = `
<div
  className="chat"
>
  John Smith
  <div
    className="messages"
  >
    <Message
      avatarUrl="/john_smith_avatar.jpg"
      isCurrentUserMessage={false}
      key="1"
      message={
        Object {
          "id": 1,
          "sender_id": 1,
          "text": "Hello, Kevin!",
        }
      }
    />
    <Message
      avatarUrl="/kevin_doe_avatar.jpg"
      isCurrentUserMessage={true}
      key="2"
      message={
        Object {
          "id": 2,
          "sender_id": 2,
          "text": "Hello, John!",
        }
      }
    />
    <Message
      avatarUrl="/john_smith_avatar.jpg"
      isCurrentUserMessage={false}
      key="1"
      message={
        Object {
          "id": 1,
          "sender_id": 1,
          "text": "How are you?",
        }
      }
    />
  </div>
  <MessageForm
    chatId={1}
  />
</div>
`;
